### DISCUSSION TITLE
How to connect to device not advertising

### DISCUSSION THREAD

**Discussion by azlekov** [JuulLabs/kable#307] [Q&A] [answered]
Hello,

How can I connect to a device which is not advertising? I saw that peripheral can be created via identifier, but is it possible to obtain identifier from a common code? I know what iOS does not expose MAC address of the devices, so do I need to save both UUID from iOS and MAC from Android to get my app working on both platforms to connect to same device?

Any clues and tactics are welcome, thanks!

**Comment by twyatt** [comment, marked as answer]
> How can I connect to a device which is not advertising?

It is my understanding that you cannot connect to a device unless it is advertising (even if you have its identifier) — at least this has been my experience on Android, not certain about the other platforms. Every time I've tried to connect to a device that isn't advertising, on Android I get a GATT 133 error.

So, at least on Android, having the identifier simply means you don't need to scan, but the device will still need to be advertising.

> I saw that peripheral can be created via identifier, but is it possible to obtain identifier from a common code?

Identifier support was added in [1.16.0](https://github.com/JuulLabs/kable/releases/tag/0.16.0).

You can retrieve the identifier via the `Peripheral.identifier` property. As you noticed, that identifier type differs based on the platform. All identifiers should work as strings. Which means you can store the identifier as a string in what ever persistent storage you use via `peripheral.identifier.toString`. To restore from your persistence, `String.toIdentifier` extension function is available.

As for creating a `Peripheral` from an identifier, that is only possible on Android and Apple (with future Javascript support coming after #286). Since it is only available on some of the platforms, it is not provided via "common" code — although, if you're not targeting Javascript, then you can provide "common" code for your project against the current API:

**Common**
```kotlin
public expect fun CoroutineScope.peripheralFromIdentifier(
    identifier: Identifier,
    builderAction: PeripheralBuilderAction = {},
)
```

**Android**
```kotlin
public actual fun CoroutineScope.peripheralFromIdentifier(
    identifier: Identifier,
    builderAction: PeripheralBuilderAction,
) = peripheral(identifier, builderAction)
```

**Apple**
```kotlin
public actual fun CoroutineScope.peripheralFromIdentifier(
    identifier: Identifier,
    builderAction: PeripheralBuilderAction,
) = peripheral(identifier, builderAction)
```

Once #286 has been resolved, then we'll be able to provide similar functions as shown above to make creating peripherals from identifiers available from "common" code.

**Reply by azlekov** [reply]
Thanks for the answer @twyatt,

> It is my understanding that you cannot connect to a device unless it is advertising (even if you have its identifier) — at least this has been my experience on Android, not certain about the other platforms. Every time I've tried to connect to a device that isn't advertising, on Android I get a GATT 133 error.

I did not explain well but you got my idea! Still, I'm thinking about something...

> Which means you can store the identifier as a string in what ever persistent storage you use via peripheral.identifier.toString. To restore from your persistence, String.toIdentifier extension function is available.

Let say I have Android and Apple app and I want to connect to same Bluetooth device from both Android and Apple device. In such case I should persist both identifiers - UUID for iOS and MAC address for Android. But in case I have connected first from Android and persist the MAC address, how I can know the UUID, so I can connect later on from the Apple device.

Any ideas how to approach this?

**Reply by twyatt** [reply]
> Let say I have Android and Apple app and I want to connect to same Bluetooth device from both Android and Apple device. In such case I should persist both identifiers - UUID for iOS and MAC address for Android. But in case I have connected first from Android and persist the MAC address, how I can know the UUID, so I can connect later on from the Apple device.

It is my understanding that all the platforms want to enforce a level of privacy for the user, hence the UUID on Apple and on newer versions of Android I believe the MAC address is randomly generated (not the actual MAC address of the peripheral you're connecting to). Because of this anonymization that both platforms are doing, I don't think you could reliably carry over an identifier to the other platform and use it for connecting.

I _think_ your best option is to have the peripheral put a unique identifier in its advertisement data — so that you can sync that data between the platforms and then scan for nearby peripherals until you find that unique advertisement data, then record the platform specific identifier. 🤷 

I can't think of an alternate approach — and my approach unfortunately means diving into your peripheral's firmware (if it doesn't already have a unique identifier in its advertisement data). 😢

**Reply by twyatt** [reply]
[This article](https://community.silabs.com/s/article/x-how-to-get-a-ble-peripheral-mac-address-with-ios-and-corebluetooth?language=en_US) discusses an approach for acquiring the MAC address on Apple (spoiler: there isn't a way via Core Bluetooth, instead your peripheral needs to provide a mechanism for retrieving it). Even if your peripheral gave you the MAC address, this would allow you to connect on Android but not Javascript (which also uses UUIDs), so it isn't a very portable/scalable solution.
